### ISSUE TITLE
hipGraph capture + MoE: numerical divergence after ~40 decoded tokens

### ISSUE THREAD

**Issue by Kaden-Schutt** [Kaden-Schutt/hipfire#19]
## Summary

When `HIPFIRE_GRAPH=1` is set for a MoE config (currently only Qwen3.5-35B-A3B, `config.num_experts > 0`), decode output starts diverging from the non-graph path around position 40+. Short decodes look fine; longer decodes degrade into token skips and degenerate loops. A safety guard in `forward_scratch` (commit `3d0bb13`) currently forces `use_graph = false` for MoE; removing it deterministically reproduces the corruption.

## Environment

- GPU: RX 7900 XTX (gfx1100, 24 GB)
- Model: `~/.hipfire/models/qwen3.5-35b-a3b.mq4` (MoE, 256 experts × top-8)
- Branch: master, tip `3d0bb13`

## Repro

Remove the `&& config.num_experts == 0` clause from `forward_scratch` (`crates/engine/src/qwen35.rs`), rebuild, run:

\`\`\`
HIPFIRE_GRAPH=1 HIPFIRE_SMOKE_MODE=chat \\
  HIPFIRE_SMOKE_PROMPT='Count from 1 to 20.' HIPFIRE_SMOKE_STEPS=100 \\
  cargo run --release --features deltanet --example a3b_smoke_forward -- \\
  ~/.hipfire/models/qwen3.5-35b-a3b.mq4
\`\`\`

## Observed

| HIPFIRE_GRAPH | Output |
|---|---|
| `0` (direct) | \`1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, ...\` correct |
| \`1\` (capture) | \`1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 9, 2, 1, 2, ...\` (skips 8, loops on 9) |

Daemon multi-turn shows the same pattern ~25 tokens into turn 2's decode: \`4 × 5 = 20\` becomes \`result is 10\` → \`2+2=2\` repeats.

## What's been audited

Every MoE-forward kernel was reviewed for graph safety. All look clean:

| Kernel | Status |
|---|---|
| \`rope_partial_interleaved_f32\` | reads pos_buf dynamically ✓ |
| \`kv_cache_write_q8_0\` | writes at pos_buf[0] ✓ |
| \`attention_flash_q8_0_tile/reduce\` | grid sized for \`max_tiles\` under capture, seq_len from pos_buf ✓ |
| \`gated_delta_net_q8\` | read-modify-write state in place, no pos kernarg ✓ |
| \`conv1d_silu_split\` | ring-buffer state updated in place ✓ |
| \`moe_softmax_topk_renorm_k8\` | single workgroup, static kernargs ✓ |
| \`gemv_hfq4g256_moe_gate_up_k8_indexed\` | static pointer + int kernargs ✓ |
| \`gemv_hfq4g256_moe_down_residual_scaled_k8_indexed\` | same ✓ |
| \`attention_q8_0_kv\` (non-flash) | has seq_len-dependent shared_mem — but capture forces \`use_flash=true\`, so this kernel is never captured |

All position-dependent values are read from device buffers (pos_buf patched per replay via \`stream_write_value32\`). All size-dependent grids/shared_mem use max-bounds under capture.

## Suspect

Most likely a subtle numerical reordering between capture-time and replay-time in one of the flash-attn tile/reduce paths or the GDN state recurrence that compounds over many replays. The divergence accumulates: the first few tokens are fine; by position ~40 the logits have drifted enough to flip argmax.

Single-turn with 30 tokens (the original smoke test that declared hipGraph \"working\") is short enough that drift hasn't accumulated yet — that's why this didn't surface initially.

## Workaround

\`forward_scratch\` in \`crates/engine/src/qwen35.rs\` gates graph capture on \`config.num_experts == 0\`. Dense Qwen3.5 still benefits from hipGraph; MoE takes the direct path. Cost: 162 → ~115 tok/s for A3B on gfx1100 (30% off the potential hipGraph peak).

## Next steps for diagnosis

Would need per-layer hidden-state dumps under both paths to isolate where they start diverging, then narrow to a specific kernel. A reference \`trace_hidden_states.rs\` example that runs the same input through both paths and diffs layer-by-layer would let the bug be bisected in a single run.

## Related commits

- \`1c69eed\` — introduced MoE hipGraph path (looked healthy at 30 tok single-turn)
- \`a05fbfa\` — first safety guard (disable graph for MoE)
- \`3d0bb13\` — expanded diagnosis + kept the guard

**Comment by nickfinease** [comment]
## Update: PR #28 does not resolve this

Reproduced on master tip (`e9c50c1`) post-PR-28 with the gfx1100 coherence fixes merged. The divergence and degenerate-attractor loop still manifest.

**Setup**
- GPU: 7900 XTX (gfx1100, 24 GB)
- Model: `~/.hipfire/models/qwen3.6-35b-a3b.mq4` (same config family as 3.5-35B-A3B per the model author)
- Branch: master @ `e9c50c1`
- Invocation: `HIPFIRE_GRAPH=1 HIPFIRE_GRAPH_MOE=1` (the existing env escape hatch — no code edit needed)
- Prompt: `Count from 1 to 20.` (chat mode, 17 prompt tokens)
- `HIPFIRE_SMOKE_STEPS=80`, `HIPFIRE_SMOKE_KV_SEQ=128`

**Graph OFF baseline (80 tokens, coherent):**
> Here's a thinking process:\n\n1. **Analyze User Input:** The user said "Count from 1 to 20."\n2. **Identify Task:** The task is straightforward: list numbers from 1 to 20 in ascending order.\n3. **D**etermine Output Format:** Just a simple list of numbers from 1 to 2...

Steady-state decode: 146.4 tok/s.

**Graph ON (80 tokens, degenerates at step 65):**
> \n\n</think>\n\nHere are the numbers from 1 to 20:\n\n1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, **14, 14, 14, 14, 14, 14, 14,**

Counts correctly through 13 (step 57), then on step 61 emits `14` a second time instead of `15`, and locks into `14,` forever. Same degenerate-attractor class as the original report's `9, 2, 1, 2, ...` — argmax has drifted far enough to fall into a numeric-token sink.

Steady-state decode: 153.4 tok/s (only ~5% above graph-OFF on this short run — capture overhead eats most of the theoretical gain at this sequence length).

**Interpretation**

The guard in `forward_scratch` is still necessary. PR #28's `__restrict__` strip, multi-block + large-LDS fix, and FP32-dequant-in-WMMA fixes all addressed dense-path kernels. The MoE-specific kernels (`moe_softmax_topk_renorm_k8`, `gemv_hfq4g256_moe_gate_up_k8_indexed`, `gemv_hfq4g256_moe_down_residual_scaled_k8_indexed`) were audited as "look clean" in the original report but weren't touched.

The next diagnostic step remains what this issue originally proposed: a `trace_hidden_states.rs` example that runs the same input through graph-OFF and graph-ON paths and diffs per-layer hidden states, so the drift can be localised to a specific MoE kernel.

Keeping open.

**Comment by Kaden-Schutt** [comment]
Resolved on master. AR-forward hipGraph is now hard-disabled (`use_graph = false`, commit `37d401e2`), which supersedes the original `num_experts == 0` guard, and the atomic-free MoE-down kernel (PR #317, `86d89537`) removed the partial root cause. With no path setting `use_graph = true`, the MoE capture divergence is unreachable. Closing — please reopen if a future re-enablement of MoE graph capture reintroduces it.

_Triaged 2026-06-03 against `origin/master`._
